ocs/ocs_agent.py

```diff
@@ -237,30 +240,62 @@ def onLeave(self, details):
         self.subscribed_topics = set()
         self.realm_joined = False
 
+    def _shutdown(self):
+        # Stop all sessions and then stop the reactor
+        self._stop_all_running_sessions()
+        try:
+            self.log.info('stopping reactor')
+            reactor.stop()
+        except ReactorNotRunning:
+            pass
+
     @inlineCallbacks
     def onDisconnect(self):
         self.log.info('transport disconnected')
+        self.log.info('waiting for reconnection')
 
         # Wait to see if we reconnect before stopping the reactor
-        timeout = 10
+        timeout = self.site_args.crossbar_timeout
+
+        # Define signal handlers to interrupt during wait for reconnection
+        def signal_handler(sig, frame):
+            self.log.info('caught {signal}!', signal=signal.Signals(sig).name)
+            self._shutdown()
+
+        signal.signal(signal.SIGINT, signal_handler)
+        signal.signal(signal.SIGTERM, signal_handler)
```

Comment:
These immediately seem odd to me, because these signal handlers will remain installed, even after a reconnection (and this function returns).  So the behavior after a reconnection is different from before the connection drop.

Can you clarify why this is needed?  Is the point that ctrl-c wasn't killing your program, or that ctrl-c wasn't running the cleanup covered in _shutdown?

If the latter, shouldn't we always have these signal handlers installed, not just if the connection has dropped out?

Reply:
Right, that's a good point that the behavior will be different. I'll come back to that.

This is needed because if a SIGINT or SIGTERM is sent to the agent during the reconnection attempt, it'll hang. For example:
```
2023-09-14T14:52:32-0400 waiting for reconnection
2023-09-14T14:52:32-0400 waiting at least 6.99999737739563 more seconds before giving up
2023-09-14T14:52:32-0400 Scheduling retry 1 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7f043ec42ad0> in 1.554297036394125 seconds.
2023-09-14T14:52:32-0400 Could not publish to Feed. TransportLost. crossbar server likely unreachable.
2023-09-14T14:52:33-0400 Scheduling retry 2 to connect <twisted.internet.endpoints.TCP4ClientEndpoint object at 0x7f043ec42ad0> in 2.826215544365843 seconds.
2023-09-14T14:52:33-0400 waiting at least 5.412115812301636 more seconds before giving up
2023-09-14T14:52:35-0400 Could not publish to Feed. TransportLost. crossbar server likely unreachable.
^C2023-09-14T14:52:35-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:40-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:40-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:40-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:41-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:41-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:41-0400 Received SIGINT, shutting down.
^C2023-09-14T14:52:54-0400 Received SIGINT, shutting down.
```

I'm trying to figure out why sending a SIGINT during normal operation (without these signal handlers in place) actually terminates this successfully, but I'm a little confused. In that case it'll get this yield and never loop:
https://github.com/simonsobs/ocs/blob/1faba0c3be3146a1bd840158ec2857cae8ccc38f/ocs/ocs_agent.py#L295

But if you send a SIGINT or SIGTERM while the reconnect loop is happening, the loop dies and its like we're stuck with the reactor running with no way out. Adding these signal handlers allows the `reactor.stop()` call to happen.

To make the behavior consistent before and after onDisconnect is called, what do you think about moving this setup to `onConnect()` or `onJoin()`?

Reply:
I've gone ahead and moved this setup to `onConnect()` in https://github.com/simonsobs/ocs/pull/337/commits/e0e73adbfb3eeec2a7c3278f4b052b6037978ede. Let me know if you'd like to handle this differently.